beacon-chain/rpc/prysm/v1alpha1/validator/duties_v2.go

```diff
@@ -0,0 +1,265 @@
+package validator
+
+import (
+	"context"
+
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/core/helpers"
+	coreTime "github.com/OffchainLabs/prysm/v6/beacon-chain/core/time"
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/core/transition"
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/rpc/core"
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/state"
+	"github.com/OffchainLabs/prysm/v6/consensus-types/primitives"
+	"github.com/OffchainLabs/prysm/v6/encoding/bytesutil"
+	"github.com/OffchainLabs/prysm/v6/monitoring/tracing/trace"
+	ethpb "github.com/OffchainLabs/prysm/v6/proto/prysm/v1alpha1"
+	"github.com/OffchainLabs/prysm/v6/time/slots"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// GetDutiesV2 returns the duties assigned to a list of validators specified
+// in the request object.
+//
+// Deprecated: The gRPC API will remain the default and fully supported through v8 (expected in 2026) but will be eventually removed in favor of REST API.
+func (vs *Server) GetDutiesV2(ctx context.Context, req *ethpb.DutiesRequest) (*ethpb.DutiesV2Response, error) {
+	if vs.SyncChecker.Syncing() {
+		return nil, status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
+	return vs.dutiesv2(ctx, req)
+}
+
+// Compute the validator duties from the head state's corresponding epoch
+// for validators public key / indices requested.
+func (vs *Server) dutiesv2(ctx context.Context, req *ethpb.DutiesRequest) (*ethpb.DutiesV2Response, error) {
+	currentEpoch := slots.ToEpoch(vs.TimeFetcher.CurrentSlot())
+	if req.Epoch > currentEpoch+1 {
+		return nil, status.Errorf(codes.Unavailable, "Request epoch %d can not be greater than next epoch %d", req.Epoch, currentEpoch+1)
+	}
+
+	// Load head state
+	s, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+	}
+
+	// Advance to start of requested epoch if necessary
+	s, err = vs.stateForEpoch(ctx, s, req.Epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	// Gather validator indices for requested pubkeys
+	requestIndices := make([]primitives.ValidatorIndex, 0, len(req.PublicKeys))
+	for _, pubKey := range req.PublicKeys {
+		if idx, ok := s.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubKey)); ok {
+			requestIndices = append(requestIndices, idx)
+		}
+	}
+
+	// Load committee and proposer metadata
+	meta, err := loadCommitteeMetadata(ctx, s, req.Epoch, requestIndices)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build duties for each validator
+	ctx, span := trace.StartSpan(ctx, "dutiesv2.BuildResponse")
```

Comment:
Why is the span so late into the function? I am not saying this is wrong, but it is unusual.

Reply:
Yeah, I added this to help understand what part of this function is taking so long